metadata_backend/api/auth.py

```diff
@@ -136,8 +136,27 @@ async def callback(self, req: Request) -> Response:
                 reason="Could not set user, missing claim CSCUserName, remoteUserIdentifier or sub."
             )
 
+        # Handle projects, they come in different formats depending on the service used.
+        # Current project sources: CSC projects, LS AAI groups
+        projects: List[str] = []
+        if "sdSubmitProjects" in session["userinfo"]:
+            # CSC projects come in format "project1 project2"
+            projects = session["userinfo"]["sdSubmitProjects"].split(" ")
+        elif "eduperson_entitlement" in session["userinfo"]:
+            # LS AAI groups come in format ["group1", "group2"]
+            groups = session["userinfo"]["eduperson_entitlement"]
+            for group in groups:
+                # LSAAI groups are of format namespaces:root_group:actual_group#client
+                # and we want to get "root_group:actual_group", and any subgroups of actual_group
+                matched = re.search(f"(?<=)({self.lsaai_root}.*?)(?=#)", group)
+                if matched:
+                    projects.append(matched.group(0))
+        else:
+            # Unknown group source, keep empty
+            pass
```

Comment:
What do we want to do if user has no projects or groups? If I recall correctly, with CSC AAI, if there are no projects, the login fails in AAI before reaching metadata-submitter, because it is configured so. But with LS AAI, do we want to let the user pass here with no groups, or raise a 400/401?

Reply:
i would raise 401, to keep it consistent with the CSC AAI, let us not introduce special cases unless we really need them.